sdk/core/az_core/src/az_credential_token.c

```diff
@@ -0,0 +1,32 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "az_credential_token_private.h"
+#include <az_platform.h>
+#include <az_spinlock_internal.h>
+
+#include <_az_cfg.h>
+
+AZ_NODISCARD az_result
+_az_credential_token_set_token(_az_credential_token* self, _az_token const* new_token)
+{
+  _az_spinlock_enter_writer(&self->_internal.lock);
+  self->_internal.token = *new_token;
+  _az_spinlock_exit_writer(&self->_internal.lock);
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result
+_az_credential_token_get_token(_az_credential_token* self, _az_token* out_token)
+{
+  _az_spinlock_enter_reader(&self->_internal.lock);
+  *out_token = self->_internal.token;
+  _az_spinlock_exit_reader(&self->_internal.lock);
+  return AZ_OK;
+}
+
+AZ_NODISCARD bool _az_token_expired(_az_token const* token)
+{
+  int64_t const expires_at_msec = token->_internal.expires_at_msec;
```

Comment:
I think reading expires_at_msec should be done under the reader lock

Reply:
We get the token, with the locks. The token includes both the token string and the expiration time. Then you can read it safely to check expiration, or to include to headers. Read az_credential_client_secret.c in this PR, lines 55-71 is how we do it.
The token does not have a lock, token credential (_az_credential_token) does. _az_token_expired takes the _az_token, not the _az_credential_token.
This way, when you reuse-or-renew-and-use token in the `_az_credential_client_secret_apply()`, you get the token, with the lock, you check expiration, and if it is good, you apply it to the header. Otherwise, you send a request to obtain the new token (without locks), then you set it to the token credential (with locks), and use the token you just obtained and set, without locks to apply to the header.
It is just one read lock operation, instead of two. And you can be sure that as you read both the token string and the expiration, they two do match each other.

Reply:
Update: to discuss this, we talked over the phone, you are fine with the idea, and asked me to update comments, and I just pushed a commit with more comments. Once CI/CD completes, I am checking this PR in. Thank you! I am resolving this conversation thread.